pkg/api/resolver.go

```diff
@@ -0,0 +1,208 @@
+package api
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	//graphql "github.com/graph-gophers/graphql-go"
+
+	"github.com/thrasher-redhat/internal-tools/pkg/db"
+)
+
+// Resolver represents the connection between the API and the data
+type Resolver struct {
+	dbClient db.Client
+}
+
+// NewResolver is a factory
+func NewResolver(db db.Client) (*Resolver, error) {
+	return &Resolver{dbClient: db}, nil
+}
+
+// parseDatestamp accounts for empty date or special strings
+func (r Resolver) parseDatestamp(datestamp string) (string, error) {
+	var date time.Time
+	var err error
+	// Default to latest
+	if datestamp == "" || datestamp == "_latest" {
+		date, err = r.dbClient.GetLatest()
+		if err != nil {
+			return "", err
+		}
+	} else if datestamp == "_earliest" {
+		date, err = r.dbClient.GetEarliest()
+		if err != nil {
+			return "", err
+		}
+	} else {
+		return datestamp, nil
+	}
+	return date.Format("2006-01-02"), nil
+}
+
+// parseComponents checks for empty lists and works around the graphql
+// listpacker bug. Basically if you use a list that isn't []interface{}, then
+// Pack() uses a interface slice where the first element is given list
+// Current workaround is to use *[]type to get the actual array, so
+// dereference that pointer here to get the actual list.
+// https://github.com/graph-gophers/graphql-go/blob/b46637030579abd312c5eea21d36845b6e9e7ca4/internal/exec/packer/packer.go#L248-L250
+// TODO - version lock graphql-go for now as the fix will break this.
+func parseComponents(components *[]string) []string {
+	// If we ant to support "_all", we could check for it in keywords and return nil
+
+	// Workaround graphql's pointer to slice bug
+	var c []string
+	if components != nil && len(*components) > 0 {
+		c = *components
+	} else {
+		c = nil
+	}
+	return c
+}
+
+// Test is a very basic example to be sure the endpoint is working
+// TODO - remove when no longer necessary
+func (r *Resolver) Test(args struct{ Text string }) string {
+	return "Hello, " + args.Text
+}
+
+// getBugs queries the database for a list of bugs and converts to BugResolvers
+func (r *Resolver) getBugs(datestamp string, components []string) ([]*BugResolver, error) {
+	// Query the database
+	bugs, err := r.dbClient.GetBugs(datestamp, components)
+	if err != nil {
+		log.Printf("Error querying client for list of bugs: %v", err)
+		return nil, err
+	}
+
+	// Convert bugs into BugResolvers
+	var brs []*BugResolver
```

Comment:
You know the length ahead of time, so do
`brs := make([]*BugResolver, len(bugs))`
and then don't use append